### ISSUE TITLE
Compound document attributes should be part of top-level included member.

### ISSUE THREAD

**Issue by m-bodmer** [SeyZ/jsonapi-serializer#8]
If I have the following config:

```javascript
JSONAPISerializerConfig = {
    resourceType: resourceType,
    topLevelLinks: {
        self: apiEndpoint
    },
    links: {
        self: function(collection) {
            return apiEndpoint + collection.id;
        }
    },
    attributes: ['title', 'slug', 'subcategories'],
    subcategories: {
        ref: 'id',
        attributes: ['id', 'title', 'description']
    }
}
```

My data object looks like:

```javascript
{
    id: 203,
    title: 'Hockey practice',
    slug: 'hockey-practice',
    subcategories: {
        description: 'Practices for strength, balance, flexibility, and vitality.',
        id: 114,
        title: 'Vitality'
    },
}
```

I get the following output:

```javascript
{
    type: "practices",
    id: "203",
    attributes: {
       title: 'Hockey practice',
    slug: 'hockey-practice',
        subcategories: {
            type: "categories",
            id: 114,
            attributes: {
               description: 'Practices for strength, balance, flexibility, and vitality.',
               id: 114,
               title: 'Vitality'
            }
        }
    }
}
```

However, my `subcategories.attributes` should exist as part of the top-level `included` member. So, it should look like the following:

```javascript
data: {
    type: "practices",
    id: "203",
    attributes: {
       title: 'Hockey practice',
    slug: 'hockey-practice',
        subcategories: {
            type: "categories",
            id: 114
        }
    }
}
included: {
  {
    "type": "categories",
    "id": "114",
    "attributes": {
      description: 'Practices for strength, balance, flexibility, and vitality.',
      id: 114,
      title: 'Vitality'
    }
  }
}
```

**Comment by SeyZ** [comment]
I cannot reproduce your problem.

Here's the code that suppose to describe your problem, right?

```
      var dataSet = {
        id: 203,
        title: 'Hockey practice',
        slug: 'hockey-practice',
        subcategories: {
          description: 'Practices for strength, balance, flexibility, and vitality.',
          id: 114,
          title: 'Vitality'
        }
      };

      new JsonApiSerializer('pratices', dataSet, {
        topLevelLinks: { self: 'http://example.com' },
        dataLinks: {
          self: function(collection) {
            return 'http://example.com/' + collection.id;
          }
        },
        attributes: ['title', 'slug', 'subcategories'],
        subcategories: {
          ref: 'id',
          attributes: ['id', 'title', 'description']
        }
      }).then(function (json) {
        console.log(require('util').inspect(json, { depth: null }));
        done(null, json);
      });
```

And here's the current result:

```
{
  data: {
    type: 'practices',
    id: '203',
    attributes: {
      title: 'Hockey practice',
      slug: 'hockey-practice'
    },
    links: {
      self: 'http://example.com/203'
    },
    relationships: {
      subcategories: {
        data: {
          type: 'subcategories',
          id: '114'
        }
      }
    }
  },
  included: [{
    type: 'subcategories',
    id: '114',
    attributes: {
      id: 114,
      title: 'Vitality',
      description: 'Practices for strength, balance, flexibility, and vitality.'
    }
  }]
}
```

Seems good, isn't it?

**Comment by m-bodmer** [comment]
Ah heres the issue:

If I put my `subcategories` object into an array, subcategories gets treated as a relationship and everything works as expected..

```javascript
subcategories: [{
    description: 'Practices for strength, balance, flexibility, and vitality.',
    id: 114,
    title: 'Vitality'
}]
```

However, does not work if I have `subcategories` like this. `subcategories` instead shows up as an attribute of `practices`

```javascript
subcategories: {
    description: 'Practices for strength, balance, flexibility, and vitality.',
    id: 114,
    title: 'Vitality'
}
```

**Comment by m-bodmer** [comment]
The problem lies here in this line of code https://github.com/SeyZ/jsonapi-serializer/blob/master/lib/serializer-utils.js#L49

`return _.isArray(obj) || _.isPlainObject(obj);`

The way I am creating the `subcategories` object from the above comment does not get returned here.

```javascript
[ { description: 'Evening yoga',
    id: 203,
    title: 'Evening yoga'} ]
isArray -> true
isPlainObject -> false
```

```javascript
{ description: 'Evening yoga',
  id: 203,
  title: 'Evening yoga' }
isArray -> false
isPlainObject -> false
```

**Comment by SeyZ** [comment]
Hmmm @m-bodmer Sorry for the missunderstanding. I need to summarize this issue.

Could you provide me:

- The dataSet that causes the problem.
- The exact current ouput.
- The exact expected output.

Thanks :)

**Comment by m-bodmer** [comment]
### Dataset

```javascript
var addressObject = [{
    address: {
        id: '54735722e16620ba1eee36af',
        addressLine1: '406 Madison Court',
        zipCode: '49426',
        country: 'USA'
    }
}];
var dataSet = [{
    id: '54735750e16638ba1eee59cb',
    firstName: 'Sandro',
    lastName: 'Munda',
    address: addressObject.address,
}, {
    id: '5490143e69e49d0c8f9fc6bc',
    firstName: 'Lawrence',
    lastName: 'Bennett',
    address: {
        id: '54735697e16624ba1eee36bf',
        addressLine1: '361 Shady Lane',
        zipCode: '23185',
        country: 'USA'
    }
}];
```

### JSON output from test suite
```javascript
{
    data: [{
        type: 'users',
        id: '54735750e16638ba1eee59cb',
        attributes: [Object]
    }, {
        type: 'users',
        id: '5490143e69e49d0c8f9fc6bc',
        attributes: [Object],
        relationships: [Object]
    }],
    included: [{
        type: 'addresses',
        id: '54735697e16624ba1eee36bf',
        attributes: [Object]
    }]
}
```

### Expected JSON output
```javascript
{
    data: [{
        type: 'users',
        id: '54735750e16638ba1eee59cb',
        attributes: [Object],
        relationships: [Object]
    }, {
        type: 'users',
        id: '5490143e69e49d0c8f9fc6bc',
        attributes: [Object],
        relationships: [Object]
    }],
    included: [{
        type: 'addresses',
        id: '54735750e16638ba1eee59cb',
        attributes: [Object]
    },{
        type: 'addresses',
        id: '54735697e16624ba1eee36bf',
        attributes: [Object]
    }]
}
```

**Comment by SeyZ** [comment]
You have the problem because your dataSet is not correct:
You should replace `address: addressObject.address,` by `address: addressObject[0].address,`.

Then, the output will be:

```
{
  data: [{
    type: 'users',
    id: '54735750e16638ba1eee59cb',
    attributes: {
      'first-name': 'Sandro',
      'last-name': 'Munda'
    },
    relationships: {
      address: {
        data: {
          type: 'addresses',
          id: '54735722e16620ba1eee36af'
        }
      }
    }
  }, {
    type: 'users',
    id: '5490143e69e49d0c8f9fc6bc',
    attributes: {
      'first-name': 'Lawrence',
      'last-name': 'Bennett'
    },
    relationships: {
      address: {
        data: {
          type: 'addresses',
          id: '54735697e16624ba1eee36bf'
        }
      }
    }
  }],
  included: [{
    type: 'addresses',
    id: '54735722e16620ba1eee36af',
    attributes: {
      'address-line1': '406 Madison Court',
      'zip-code': '49426',
      country: 'USA'
    }
  }, {
    type: 'addresses',
    id: '54735697e16624ba1eee36bf',
    attributes: {
      'address-line1': '361 Shady Lane',
      'zip-code': '23185',
      country: 'USA'
    }
  }]
}
```

**Comment by estefaniaa** [comment]
@m-bodmer, @SeyZ I have the same issue, but with a single resource (ex. `/collections/209`)

## Dataset
```
{ 
  id: 209,
  title: 'My Collection',
  subcategory:
   { 
     title: 'My Subcategory'
     id: 119
   } 
}
```

## Current output
```
{
  "data": {
    "type": "collections",
    "id": "209",
    "attributes": {
      "title": "My Collection",
      "subcategory": {
        "id": 119,
        "title": "My Subcategory"
      }
    }
  },
  "links": {
    "self": "http://localhost:9642/api/v4/collections/209"
  }
}
```

## Desired output
```
{
  "data": {
    "type": "collections",
    "id": "209",
    "attributes": {
      "title": "My Collection"
    },
    "relationships": {
      "subcategory": {
        "data": {
            "type": "subcategories",
            "id": "119"
        },
        "links": {
          "self": "http://localhost:9642/api/v4/collections/209/relationships/subcategory"
        }
      }
    }
  },
  "links": {
    "self": "http://localhost:9642/api/v4/collections/209"
  },
  "included": [
    {
      "type": "subcategories",
      "id": "119",
      "attributes": {
        "title": "My Subcategory",
      },
      "links": {
        "self": "http://localhost:9642/api/v4/subcategories/119"
      }
    }
  ]
}
```

If I wrap the single `subcategory` object in an array, then it returns the correct payload in `included`, but then `relationships.subcategory.data` contains an array which is not desired as I want to preserve the one-to-one relationship.

**Comment by SeyZ** [comment]
@estefaniaa Could you provide me one more thing: the `JsonApiSerializer` call?
Example:

```
new JsonApiSerializer('users', dataSet, {
  attributes: ['title', 'subcategory'],
  subcategory: {
    ref: 'id',
    attributes: ['title']
  }
}).then(function (json) {
  console.log(require('util').inspect(json, { depth: null }));
}
```

**Comment by estefaniaa** [comment]
@SeyZ, sure, here's the config -- looks pretty much the same as yours:

```
var jsonApiSerializerConfig = {
	links: {
		self: 'http://localhost:3000/collections/' + collection.id
	},
	attributes: ['title', 'subcategory'],
	subcategory: {
		ref: 'id',
		attributes: ['title'],
		includedLinks: {
			self: function (subcategory) {
				return 'http://localhost:3000/subcategories/' + subcategory.id;
			}
		},
		relationshipLinks: {
			self: function (subcategory) {
				return http://localhost:3000/collections/' + collection.id + '/relationships/subcategory';
			}
		}
	}
};
```

**Comment by estefaniaa** [comment]
@SeyZ @m-bodmer I've solved the issue on my end -- the `subcategory` object being passed is a query result from a [Waterline](https://github.com/balderdashy/waterline) model. While the dataset, when logged, is as I described above, it is a pretty-print/toString of the actual Waterline model object, which appears to not be easily parseable by the serializer.

To solve, I simply defined a [`toJSON()`](https://github.com/balderdashy/waterline#model) method in my model and invoke that method before passing the dataset to the serializer.
